Offline testing

jsumners opened the pull request:
This pull request re-works the whole test suite to use a mock server. Thus, the tests are able to be run without a hard dependency on any individual Active Directory server.

It also fixes a couple bugs that were shown as a result of the updates. Mainly that the default attributes were out-of-sync with what the documentation said they were. To alleviate this, I expose the `defaultAttributes` as a property on the `ActiveDirectory` function.

Further, I re-named the test suites to match their script names. This means the following commands are available for individual testing:
- `mocha --ui bdd --grep ^Attributes\s`
- `mocha --ui bdd --grep ^Authentication\s`
- `mocha --ui bdd --grep ^Network\s`
- `mocha --ui bdd --grep ^ctor\s`
- `mocha --ui bdd --grep ^find\s`
- `mocha --ui bdd --grep ^findGroup\s`
- `mocha --ui bdd --grep ^findGroups\s`
- `mocha --ui bdd --grep ^findUser\s`
- `mocha --ui bdd --grep ^findUsers\s`
- `mocha --ui bdd --grep ^getGroupMembershipForGroup\s`
- `mocha --ui bdd --grep ^getGroupMembershipForUser\s`
- `mocha --ui bdd --grep ^getUsersForGroup\s`
- `mocha --ui bdd --grep ^groupExists\s`
- `mocha --ui bdd --grep ^isUserMemberOf\s`
- `mocha --ui bdd --grep ^userExists\s`

I also switched the assertion library to Chai and updated all of the test code to ES2015 syntax. So the `assertmore` script got removed and the the underscore library is no longer used (in the tests).

jsumners committed:
Add testing script to package.json

jsumners committed:
Update network tests

This commit updates the network.js tests to use chai assertions.
It also fixes a bug wherein the callback was not being invoked on
error during authentication.

jsumners committed:
Add mock AD server and update authenticate tests

This commit adds a mock LDAP server that emulates some Active
Directory functionality -- non-conforming usernames. The
authenticate.js tests were updated to use this mock server
and the chai assertion library.

jsumners committed:
Update attributes testing

jsumners committed:
Update userExists tests

jsumners committed:
Update groupExists tests

jsumners committed:
Updates getGroupMembershipForUser tests and fixes bug

This commit updates the getGroupMembershipForUser tests.
It also fixes a bug wherein the documented default attributes
for a group query were not reflected in the code.

jsumners committed:
Update isUserMemberOf tests

jsumners committed:
Update getUsersForGroup tests and fix default user attributes bug

This commit updates the getUsersForGroup tests. It also fixes a
bug wherein the documented default user attributes were not being
returned correctly.

jsumners committed:
Update ctor tests

jsumners committed:
Update find method tests

jsumners committed:
Update findGroup method tests

jsumners committed:
Fix findGroups bug

This commit fixes a bug in #findGroups() wherein it was "picking"
the default attributes for users, not groups.

jsumners committed:
Update findGroups tests

jsumners committed:
Use exported default attributes in tests

jsumners committed:
Update findUser tests

jsumners committed:
Remove unused requires from previous commit

jsumners committed:
Update findUsers method tests

jsumners committed:
Update getGroupMembershipForGroup method tests

jsumners committed:
Fix "Director*" findGroups test

jsumners committed:
Fixes so all tests run instead of only by group

jsumners committed:
Fix "AnInvalidUsername" test

This commit adds support for returning the correct error
type when the AnInvalidUsername test is invoked. Previously,
it was a crapshoot as to when the test would pass. And if it
didn't pass then random other tests would fail.

gheeres commented:
Thanks for the pull request. 

Please give me some time to review the changes. A quick glance looks **GREAT**!

jsumners commented:
No problem. It's a lot to review.

jsumners commented on test/mockServer/server.js:
```diff
@@ -0,0 +1,5 @@
+'use strict';
```
This file was used so that I could issue queries against it with ldapsearch. It should probably be removed.

jsumners committed:
Add comments and clean up a few things

This commit attempts to clarify the mockServer implementation
with some more comments. It also cleans up a couple items such
as unused variables and dead code/comments.

jsumners committed:
Merge upstream master

This commit merges in upstream master to remedy
merge conflicts.

jsumners commented:
I have updated my branch with some clarification comments and elimination of merge conflicts.

jsumners committed:
Fix ctor#InvalidCredentialsError test
